File: ocaml/lib/kan/lift.ml
module Lift = struct
  type ('g, 'h, 'a) t = ('a -> 'b) -> 'g 'a -> 'h 'b

  let dimap (f : 'b -> 'a) (g : 'a -> 'c) (m : ('g, 'h, 'a) t) : ('g, 'h, 'c) t =
    fun bac r ->
      m (fun a -> bac (g a)) (f r)

  let divide (f : 'a -> 'b) (g : ('g, 'h, 'a) t) (h : ('g, 'h, 'b) t) : ('g, 'h, 'c) t =
    fun bac r ->
      h (fun b -> bac b) (g (fun a -> bac (f a)) r)
end

File: rust/src/kan/lift.rs
use std::ops::Fn;

pub struct Lift<G, H, A>
where
    G: Clone,
{
    value: Box<dyn Fn(Box<dyn Fn(A) -> B>, G) -> H>,
}

impl<G, H, A, B, C> Lift<G, H, A>
where
    G: Clone,
{
    pub fn dimap<F, FG>(
        &self,
        f: F,
        g: impl Fn(A) -> C,
    ) -> Lift<G, H, C>
    where
        F: Fn(B) -> A + 'static,
        FG: Fn(G) -> G + 'static,
    {
        Lift {
            value: Box::new(move |bac, a| {
                self.value(Box::new(move |a| bac(g(a))), FG(a.clone()))
            }),
        }
    }

    pub fn divide<B>(&self, f: impl Fn(A) -> B + 'static, s: Lift<G, H, B>) -> Lift<G, H, C>
    where
        B: 'static,
        C: 'static,
    {
        Lift {
            value: Box::new(move |bac, a| {
                s.value(Box::new(move |b| bac(b)), self.value(Box::new(move |a| bac(f(a))), a.clone()))
            }),
        }
    }
}